src/dart/ast.rs

```diff
@@ -0,0 +1,479 @@
+            BinOp::Bool(BoolBinOp::Ge) => ">=",
+            BinOp::Bool(BoolBinOp::Gt) => ">",
+            BinOp::Bool(BoolBinOp::Le) => "<=",
+            BinOp::Bool(BoolBinOp::Lt) => "<",
+
+            BinOp::Value(op) => value_bin_op!(op, ""),
+
+            BinOp::Assign(Some(op)) => value_bin_op!(op, "="),
+            BinOp::Assign(None) => "=",
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum Type {
+    Path(Qualified, Vec<Box<Type>>),
+    FunctionOld(FnSig),
+    Function(FnSig),
+    Infer,
+}
+
+#[derive(Debug)]
+pub enum Expr {
+    Unary(UnOp, Box<Expr>),
+    Binary(BinOp, Box<Expr>, Box<Expr>),
+    Conditional(Box<Expr>, Box<Expr>, Box<Expr>),
+    Is(Box<Expr>, Box<Type>),
+    IsNot(Box<Expr>, Box<Type>),
+    As(Box<Expr>, Box<Type>),
+    Suffix(Box<Expr>, Suffix),
+    Identifier(Symbol),
+    FunctionExpr(FnArgs, FunctionBody),
+    New {
+        const_: bool,
+        ty: Box<Type>,
+        ident: Option<Symbol>,
+        args: Arguments,
+    },
+    List {
+        const_: bool,
+        element_ty: Option<Box<Type>>,
+        elements: Vec<Box<Expr>>,
+    },
+    Map {
+        const_: bool,
+        kv_ty: Option<(Box<Type>, Box<Type>)>,
+        kv: Vec<(Box<Expr>, Box<Expr>)>,
+    },
+    Number(Symbol),
+    String(Vec<StringLiteral>),
+    Paren(Box<Expr>),
+    Symbol(SymbolLiteral),
+    Throw(Box<Expr>),
+    Cascade(Box<Expr>, Cascade),
+}
+
+#[derive(Debug)]
+pub enum SymbolLiteral {
+    Operator(OverloadedOp),
+    Path(Vec<Symbol>),
+}
+
+#[derive(Debug)]
+pub struct StringLiteral {
+    pub raw: bool,
+    pub triple: bool,
+    pub quote: char,
+    pub prefix: Symbol,
+    pub interpolated: Vec<(Box<Expr>, Symbol)>,
+}
+
+#[derive(Debug)]
+pub enum Suffix {
+    Index(Box<Expr>),
+    Field(Symbol),
+    FieldIfNotNull(Symbol),
+    Call(Vec<Box<Type>>, Arguments),
+}
+
+#[derive(Debug)]
+pub struct Cascade {
+    pub suffixes: Vec<Suffix>,
+    pub assign: Option<(Option<ValueBinOp>, Box<Expr>)>,
+}
+
+#[derive(Debug)]
+pub enum UnOp {
+    Neg,
+    Not,
+    Tilde,
+    Await,
+    PreInc,
+    PreDec,
+    PostInc,
+    PostDec,
+}
+
+#[derive(Debug)]
+pub struct NamedArgument {
+    pub name: Symbol,
+    pub expression: Box<Expr>,
+}
+
+#[derive(Debug)]
+pub struct Arguments {
+    pub unnamed: Vec<Box<Expr>>,
+    pub named: Vec<NamedArgument>,
+}
+
+#[derive(Debug)]
+pub struct MetadataItem {
+    pub qualified: Qualified,
+    pub suffix: Option<Symbol>,
+    pub arguments: Option<Arguments>,
+}
+
+pub type Metadata = Vec<MetadataItem>;
+
+#[derive(Debug)]
+pub enum FinalConstVar {
+    Final,
+    Const,
+    Var,
+}
+
+#[derive(Debug)]
+pub struct FnSig {
+    pub return_type: Box<Type>,
+    pub args: FnArgs,
+}
+
+#[derive(Debug)]
+pub enum OptionalFormalParameter {
+    Positional,
+    Named,
+}
+
+impl Default for OptionalFormalParameter {
+    fn default() -> Self {
+        OptionalFormalParameter::Positional
+    }
+}
+
+#[derive(Default, Debug)]
+pub struct FnArgs {
```

Comment:
I'd move all the fields into `FnSig` and use that (with `Type::Infer` for the return type when that isn't applicable).